### PR TITLE
Block production trigger

### PR DESCRIPTION
**Pull request by Dentosal**
Closes https://github.com/FuelLabs/fuel-core/issues/50

~~Requires #409 to be merged first~~

### CODE REVIEW

**Comment by Voxelot** [comment]
FYI:

I retargeted this PR to diff against #409 

As far as configuration is concerned, we'll want a way to connect/consolidate with the `ProductionStrategy` chainspec configuration enum: https://github.com/FuelLabs/fuel-core/blob/v0.10.1/fuel-core/src/chain_config.rs#L107

**Comment by Dentosal** [comment]
Ready for the initial round of comments

File: fuel-block-producer/src/block_producer/transaction_selector.rs
Comment (Voxelot) [inline fuel-block-producer/src/block_producer/transaction_selector.rs:1]: This file was moved to the TxPool adapter, as this logic should eventually move to txpool itself. I think this may have re-appeared due to a merge conflict.

**Comment by bvrooman** [comment]
Can we upgrade tokio in a separate PR to reduce some of the noise? It would remove a handful of files from your PR.

File: fuel-poa-coordinator/src/service.rs
Comment (bvrooman) [inline fuel-poa-coordinator/src/service.rs:186]: nit: Just for the sake of consistency: 

```suggestion
                let block_tx_count =
                    self.signal_produce_block().await?.transactions.len();
                let includable_txs_count = self.txpool_sender.includable().await?.len();
```

File: fuel-poa-coordinator/src/service.rs
Comment (bvrooman) [inline fuel-poa-coordinator/src/service.rs:225]: Same as above:

```suggestion
                    let block_tx_count =
                        self.signal_produce_block().await?.transactions.len();
                    let includable_txs_count = self.txpool_sender.includable().await?.len();
```

File: fuel-poa-coordinator/src/service.rs
Comment (bvrooman) [reply fuel-poa-coordinator/src/service.rs:186]: Also, `includable()` returns the sorted txs. If you're only getting the count of the txs in the tx pool, and the order doesn't matter, you could get away with some sort of `size()` call on the `txpool_sender`, but that currently doesn't exist.

I would suggest maybe investigating to see if we can do a similar optimization to get the upcoming size of the block from the block producer without going through all the production steps.

File: fuel-block-producer/src/block_producer/transaction_selector.rs
Comment (ControlCplusControlV) [inline fuel-block-producer/src/block_producer/transaction_selector.rs:25]: Should we have some method to seperate `select_transaction` strategies? So either renaming it, or using some shared interface could be helpful. So that block producers can choose between `select_tx_gas_sorted` and `select_mev_tx_priority` etc, etc

File: fuel-block-producer/src/config.rs
Comment (ControlCplusControlV) [inline fuel-block-producer/src/config.rs:15]: Was this TODO fixed / survived somewhere else?

File: fuel-core/src/service/config.rs
Comment (ControlCplusControlV) [inline fuel-core/src/service/config.rs:73]: [nit] should `networksc` be `networksc`? I learned `iff` was a thing earlier this week so not sure if this is some abbreviation, if it is would be nice to be more clear

File: fuel-core/src/service/modules.rs
Comment (ControlCplusControlV) [reply fuel-core/src/service/modules.rs:143]: If not spinning up a separate thread to just receive like

```
tokio::spawn(async move { while (tx_status_receiver.recv().await).is_ok() {} });
```

With a TODO or issue to remove it somewhere

File: fuel-core/src/service/modules.rs
Comment (ControlCplusControlV) [inline fuel-core/src/service/modules.rs:170]: Is there a more clear way we can distinguish BFT is disabled because node is currently in PoA configuration

**Comment by Dentosal** [comment]
See https://github.com/FuelLabs/fuel-core/pull/662 for tokio upgrade to simplify this one

File: fuel-block-producer/src/config.rs
Comment (Dentosal) [reply fuel-block-producer/src/config.rs:15]: It was moved to be an input argument to the trigger, see fuel-poa-coordinator/src/config.rs

File: fuel-core/src/service/modules.rs
Comment (Dentosal) [reply fuel-core/src/service/modules.rs]: No, my rust-analyzer sometimes adds those without any need to. Removed.

File: fuel-poa-coordinator/src/service.rs
Comment (Dentosal) [reply fuel-poa-coordinator/src/service.rs]: On thxpool side, we are sending a vector of `Arc`'d transactions and then taking count. The complier might be able to optimize this to just sending then len over.

File: fuel-block-producer/src/block_producer/transaction_selector.rs
Comment (Dentosal) [reply fuel-block-producer/src/block_producer/transaction_selector.rs]: This is outside the scope of this PR. Maybe you could open a new issue if that needs discussion?

File: fuel-block-producer/src/block_producer/transaction_selector.rs
Comment (Dentosal) [reply fuel-block-producer/src/block_producer/transaction_selector.rs]: Removed.

File: fuel-core/src/service/modules.rs
Comment (ControlCplusControlV) [reply fuel-core/src/service/modules.rs:174]: Just include in the commeent as to why `bft.sender().clone()` is commented out rather than not outright deleted

Comment (Dentosal) [reply fuel-core/src/service/modules.rs:174]: Sure

**Comment by Voxelot** [comment]
Looks like we'll need a separate PR for enabling the real block producer in fuel-core and disabling the `executor.submit_txs` flow so that this poa coordinator is used instead.

**Comment by Voxelot** [comment]
Also, have you tried running fuel-core via CLI to see what happens when txs are submitted? I'm not seeing any options on the CLI to configure the block production settings.

**Comment by Dentosal** [comment]
I've added the production settings to ChainConfig. I'm not sure if they should instead be command line parameters for fuel-core instead, but chain config seems potentially better place to keep this.

**Comment by Voxelot** [comment]
ChainConfig does make sense. The reason I was thinking of having CLI options to select the mode is that during development or debugging. It's a fairly common practice in Ethereum to fork the mainnet chain and then put it into instant mode for local testing. 

Given it's not that hard to edit the chainconfig json, I don't mind leaving the settings there for now.

File: fuel-core/src/cli/run.rs
Comment (Voxelot) [inline fuel-core/src/cli/run.rs:130]:
```
            block_producer: fuel_block_producer::Config {
                consensus_params: chain_conf.transaction_parameters
            }
```

**Comment by Dentosal** [comment]
This has some weird timeout issues after merging the latest master, as seen by the failed test run. Introduced here in commit 62b2826c2a4af257b84311dbfb3ad2fc48a14b82. I'll try to locate the offending change tomorrow.

File: fuel-poa-coordinator/src/service.rs
Comment (ControlCplusControlV) [inline fuel-poa-coordinator/src/service.rs:319]: Error processing I think should be fleshed out more here. If `process_next_event` errs with a txpool error it might be reasonable to dump the txpool and try to re-populate it, while if there is an error with block production skipping the block or some other recourse like changing block production strategy may be a more desireable outcome. So maybe seperate out the err cases within here to catch both of those, then as a final exhaustive match panic

File: fuel-poa-coordinator/README.md
Comment (ControlCplusControlV) [inline fuel-poa-coordinator/README.md:3]: Do we want PoS references in READMEs?

**Review by ControlCplusControlV — DISMISSED** [review]
LGTM, don't see any block changes still. Await feedback from the others before merging though

**Comment by freesig** [comment]
I'm getting this error when I try to run the tests in `fuel-poa-coordinator`:
image

File: fuel-block-producer/src/block_producer/tests.rs
Comment (freesig) [inline fuel-block-producer/src/block_producer/tests.rs:56]: Seems like there should be a new test in that checks what happens if the included txs are > `max_gas`? Or at least showing that > `max_gas` txns never gets included. Or maybe it's just the wrong level to test that?

**Review by freesig — CHANGES_REQUESTED** [review]
I'd like to be able to run the poa-coordinator tests before approving this. It appears they aren't running in CI as CI is passing?

**Comment by Dentosal** [comment]
Waiting for https://github.com/FuelLabs/fuel-core/pull/688

File: fuel-poa-coordinator/README.md
Comment (Voxelot) [reply fuel-poa-coordinator/README.md:3]: There could also be scenarios where a PoA configuration is desired in the future for testing or other use-cases. So we should describe what this module accomplishes in a generic sense independently of our roadmap.

File: fuel-poa-coordinator/src/service.rs
Comment (Voxelot) [inline fuel-poa-coordinator/src/service.rs:186]: Since we are moving the selection of transactions to the txpool, we shouldn't need to use the block producer at all here.

Following from the P&A architecture proposal, the POA module could define it's own trait for getting exactly what it wants from the txpool. Since the POA coordinator only caree about whether total max gas consumable by all the txs in the pool is > the max gas per block, it could have a port like:

```
trait TransactionPool {
    fn get_total_comsumable_gas(&self) -> u64;
}
```

And then use it here to check: `self.txpool.get_total_comsumable_gas() > config.max_gas_per_block()`.

This way, the Poa doesn't risk producing extra or empty blocks for no reason.

File: fuel-poa-coordinator/src/service.rs
Comment (Voxelot) [inline fuel-poa-coordinator/src/service.rs:227]: See comment above - also this logical check on the txpool size vs block capacity should be deduplicated into a different function.

File: fuel-poa-coordinator/Cargo.toml
Comment (Voxelot) [inline fuel-poa-coordinator/Cargo.toml:24]: We should avoid depending on any other workspace members in these domain crates except for maybe `fuel-core-interfaces`.

Integration tests that fully depend on big dependencies like this should only exist in `fuel-tests`:

- This gives us assurances that the component actually works inside a real instance of fuel-core.
- The harness in fuel-tests is optimized to reduce the number of targets big dependencies like rocksdb must be linked into
- Changes to other crates like block producer or fuel-core won't drastically impact the structure of the domain-focused tests.


From what I can tell, the only reason we need fuel-core is to provide an impl of InterpreterStorage which has a method for returning the block height. However, the POA module shouldn't need to care about any of the other trait requirements of interpreter storage, as it never interacts with an instance of the VM.

The PoA orchestrator should have its own database trait for fetching the block height, and use a mock impl for testing instead of depending on fuel-core's implementation.

File: fuel-poa-coordinator/src/service.rs
Comment (Voxelot) [inline fuel-poa-coordinator/src/service.rs:68]: See comment above: https://github.com/FuelLabs/fuel-core/pull/599/files#r991677314

It's better to create traits with similar getters, than to reuse traits with many other useless things (from a PoA perspective) that make mocking difficult here.

File: fuel-poa-coordinator/src/service.rs
Comment (Dentosal) [reply fuel-poa-coordinator/src/service.rs:319]: I'm not exactly sure what do you mean by repopulating the txpool here? I don't think we even do any fallible txpool calls anyway. My reasoning is that if we end up taking the error path here, it must already be a programming error somewhere. I would like to panic early instead of doing this, since it's hightly unlikely that any error can be recovered from.

File: fuel-block-producer/src/block_producer/tests.rs
Comment (Dentosal) [reply fuel-block-producer/src/block_producer/tests.rs:56]: I think this is out of scope for this PR, but such tests should be written anyway. I'll create an issue when this PR gets merged.

File: fuel-poa-coordinator/Cargo.toml
Comment (Dentosal) [reply fuel-poa-coordinator/Cargo.toml]: Removed fuel-core dev-dependency and factored block height query to a separate trait.  The block producer dependency is removed as well, by simply copying the imported function as-is, which is actually nice as the test of poa-coordinator shouldn't depend on internals of the block producer.

File: fuel-poa-coordinator/src/service.rs
Comment (Dentosal) [reply fuel-poa-coordinator/src/service.rs]: Done

File: fuel-poa-coordinator/src/service.rs
Comment (Dentosal) [reply fuel-poa-coordinator/src/service.rs:222]: Noted, continuing discussion in the item above

File: fuel-poa-coordinator/src/service.rs
Comment (Dentosal) [reply fuel-poa-coordinator/src/service.rs:181]: I now have local version that does exactly this. However, introducing additional channels makes the test timings even harder, and now some of the test cases are flaky, The flakyness is rather hard to work around, for the fix see this tokio issue https://github.com/tokio-rs/tokio/issues/4522  (and also tokio PR 4523, for an earlier attempt to fix this)

**Comment by Voxelot** [comment]
clippy is failing

File: fuel-poa-coordinator/src/deadline_clock.rs
Comment (Voxelot) [inline fuel-poa-coordinator/src/deadline_clock.rs:132]: [nit]
```suggestion
    /// Sets the timeout, optionally overwriting the existing value
```

File: fuel-poa-coordinator/src/deadline_clock.rs
Comment (Voxelot) [inline fuel-poa-coordinator/src/deadline_clock.rs:121]: [nit]

```suggestion
    /// Sets the timeout, optionally overwriting the existing value
```

File: fuel-core/src/service/modules.rs
Comment (Voxelot) [inline fuel-core/src/service/modules.rs:91]:
```suggestion
        // TODO: enable when bft config is ready to use
```

File: fuel-poa-coordinator/src/service.rs
Comment (Voxelot) [inline fuel-poa-coordinator/src/service.rs:223]: [nit] Should we move this to an else clause below and only update the timer if we don't immediately jump into block production?

Comment (Dentosal) [reply fuel-poa-coordinator/src/service.rs]: Makes sense

File: fuel-poa-coordinator/tests/test_trigger.rs
Comment (Voxelot) [inline fuel-poa-coordinator/tests/test_trigger.rs:680]: [nit] 
```suggestion
async fn hybrid_trigger_reacts_correctly_to_full_txpool() -> anyhow::Result<()> {
```

**Review by Voxelot — DISMISSED** [review]
very cool use of tokio time!

the hybrid trigger is the most complicated piece here and also has the least amount of tests. But we can cover this more extensively during the integration with fuel-core.

The only thing that may need to change based on the current arch discussion is that notifying the txpool about the new block should probably be done in the block importer. But since the importer doesn't exist yet, we can defer that.

**Review by Voxelot — APPROVED** [review]

File: fuel-block-producer/src/block_producer/tests.rs
Comment (Dentosal) [reply fuel-block-producer/src/block_producer/tests.rs:56]: https://github.com/FuelLabs/fuel-core/issues/695

**Pull request merged**
